### ISSUE TITLE
DAOS: tcp;ofi_rxm server segfault during failed connection attempt by the client

### ISSUE THREAD

**Issue by frostedcmos** [ofiwg/libfabric#7417]
Noticed during the periodic run of CI, one of the DAOS tests sporadically failed. In the test, servers and clients are started, with client attempting to send 1 RPC to each server.

The client end reported a timeout attempting to connect to one of the servers (60 second no response). The server then crashed with the following stack trace:

Thread 1 (Thread 0x7f7681fb5700 (LWP 67525)):
#0  0x00007f76bce9d5bb in rxm_process_connect (cm_entry=cm_entry@entry=0x7f7681fb2e90) at prov/rxm/src/rxm_conn.c:409
        conn = 0x7f76bd14cd40 <tcpx_msg_ops>
#1  0x00007f76bce9d710 in rxm_handle_event (ep=ep@entry=0x7f768c033d90, event=2, cm_entry=cm_entry@entry=0x7f7681fb2e90, len=16) at prov/rxm/src/rxm_conn.c:678
        __func__ = "rxm_handle_event"
        saved_errno = <optimized out>
#2  0x00007f76bce9e1ff in rxm_cm_progress (arg=0x7f768c033d90) at prov/rxm/src/rxm_conn.c:776
        ep = 0x7f768c033d90
        cm_entry = {fid = 0x7f768c215fd0, info = 0x0, data = {connect = {version = 1 '\001', endianness = 2 '\002', ctrl_version = 4 '\004', op_version = 3 '\003', port = 31417, padding = "\000", eager_limit = 16384, rx_size = 4096, client_conn_id = 8}, accept = {server_conn_id = 134935038132737, rx_size = 16384}, reject = {version = 1 '\001', reason = 2 '\002'}}}
        event = 2
        ret = 16
        __func__ = "rxm_cm_progress"
#3  0x00007f76c371a14a in start_thread () from /lib64/libpthread.so.0
No symbol table info available.
#4  0x00007f76c2947f23 in clone () from /lib64/libc.so.6
No symbol table info available.
No core file now.

Corresponding DAOS ticket https://daosio.atlassian.net/browse/DAOS-9786

**Comment by shefty** [comment]
Can you re-run with the latest libfabric main and let me know if you still see the segfault?  There were 3 recent patches added, one of which addressed a segfault issue with rxm over tcp.  (I'm skeptical that those fixes are related to this crash, however).

I'm assuming this is with v1.14.x.  If so, the segfault is pointing to:
```
409	conn->ep->connecting_cnt--;
410	assert(conn->ep->connecting_cnt >= 0);
```

**Comment by frostedcmos** [comment]
Yes this was with v1.14.x; we will retry with latest ofi main and let you know.

**Comment by shefty** [comment]
Free build or debug build?

**Comment by frostedcmos** [comment]
this was hit in our CI which runs debug builds. unfortunately there isnt an exact sequence to reproduce this; we seem to hit this very rarely/sporadically in CI once in a while. I will try some experiments locally to get more consistent reproducer

**Comment by shefty** [comment]
Knowing it was a debug build helps.  This is the surrounding code:
```
	assert(ofi_ep_lock_held(&conn->ep->util_ep));
	if (conn->state == RXM_CM_CONNECTING)
		conn->remote_index = cm_entry->data.accept.server_conn_id;

	conn->ep->connecting_cnt--;
	assert(conn->ep->connecting_cnt >= 0);
```
I was wondering if conn->ep was invalid, but since this is a debug build, the assert higher up would have already dereferenced that pointer.  So, I think the assert for connecting_cnt was hit, which suggests a race somewhere in connection event handling.

**Comment by shefty** [comment]
Please let me know if you have hit this same assert at least one other time.  I've been analyzing the code paths around this assert, looking for a race somewhere, and so far I haven't been able to find one.  I'm guessing there is a race here I'm not seeing.

**Comment by shefty** [comment]
After analyzing the code, my best guess as to what might be triggering this is that the tcp provider is reporting an event for an endpoint that has been closed.  That could conceivably cause all sorts of issues with rxm trying to access freed memory.  I can't say that this is happening for certain, but the cleanup code for destroying an endpoint has gaps in the locking that make me question that there's proper synchronization.  I've restructured the cleanup path to eliminate the gaps in PR #7457.
